docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,534 @@
+rrDHT is designed with a few performance requirements/characteristics in mind.
+
+1. It must have a compact and computationally simple representational model for identifying which nodes are responsible for which content, and which nodes actually hold which content. (A **"world model"** of what is where.)
+2. It must have **lookup speeds** at least as fast as Kademlia's binary trees ( $\mathcal{O}(n \log n)$ ). Current testing shows an average of 3 hops/queries to reach an authority with the data.
+3. It must be **adjustable** to be both resilient and performant across many DHT compositional make-ups (reliability of nodes, different network topologies, high/low usage volumes, etc.)
+
+**World Model:** The network location space is a circle comprising of the range of unsigned 32-bit numbers, in which the location $0$ is adjacent to the location ${2^{32}}-1$.
+
+$$
+\begin{aligned}
+L &: \{l_0,\dots,l_{2^{32}-1}\} \\
+l_{2^{32}-1} &< l_0
+\end{aligned}
+$$
+
+The larger 256-bit address space of the DHT, consisting of 256-bit "basis hashes" $B$ (Blake2b-256 hashes of addressable content $C$, and Ed25519 public keys of agents $K$), is mapped to the smaller network location space via a function $\mathsf{map\_to\_loc}(b)$, which is the XOR of 8 × 32-bit segments of the hashes. The original address is still kept with the data or metadata, so collisions in the smaller space are not a concern.
+
+$$
+\begin{aligned}
+E &: \{e_0,\dots,e_{2^{256}-1}\} \\
+K &: \{k_0,\dots,k_{2^{256}-1}\} \\
+C &: E \cup K \\
+\mathsf{hash} &: C \rightarrow B \\
+\mathsf{map\_to\_loc} &: B \rightarrow L \\
+|L| &\approx |B| \cdot 2^8
+\end{aligned}
+$$
+
+Each agent has a network location $l_k$ in this 32-bit space as well as an arc size $s_{arc}$ indicating how large an 'arc' of the location circle they are claiming authority for. The storage arc $ARC_{l_k}$ defines the range of basis hashes for which a node claims authority. The arc spreads clockwise from $l_k$.
+
+$$
+ARC_{l_k} : \{l_k,\dots,l_k + s_{arc}\}
+$$
+
+This guarantees that a node can rapidly resolve any basis hash $b$ to the most likely candidate for an authority $k_{best}$ by comparing the basis hash's network location $l_b$ to all the authorities they know about in the set $L_K$ using the following algorithm, in pseudocode:
+
+```
+k_best = L_K
+    .sort_ascending()
+    .filter(l_k -> l_k <= l_b)
+    .last()
+```
+
+#### Network Location Quantization
+
+Additionally, arcs can be subjected to **quantization** which splits the network location space $L$ into disjoint subsets of a given quantum size $s_q$, and to which the starting arc boundary $k$ and arc size $s_{arc}$ are also snapped. The quantized arc size is then represented as an 8-bit integers containing the current size of quantum chunks of the address space and the number of chunks in the arc. Peers can also quantize their held operations in the time dimension using a quantum size that exponentially increases as the dimension extends into the past.
```

Comment:
Needed some clearing up. This might be a good place to link to an appendix, too. Note that the time chunk size increases quadratically, not exponentially.

```suggestion
Additionally, arcs are be subjected to **quantization** which splits the network location space $L$ into disjoint subsets of a given size $s_q$, and to which the starting arc boundary $k$ and arc size $s_{arc}$ are also snapped. The quantized arc is then fully represented by three numbers: the quantized chunk size $s_q$, the number of chunks until the start boundary $k_q$, and the number of chunks from start to end $n_q$.


Peers also quantize the time dimension such that the size of chunks of time increase quadratically as the dimension extends into the past.
```